### ISSUE TITLE
voicebox-server memory grows ~unbounded over consecutive generations (CPU backend); once large, new generations hang in "generating" forever

### ISSUE THREAD

**Issue by Rashinban1988** [jamiepine/voicebox#923]
## Environment

- Voicebox 0.5.0 (Windows MSI install)
- Windows 11 [10.0.26200], CPU backend (`backend_variant: cpu`, `gpu_available: false`)
- Machine: Intel Core Ultra 9 288V (Arc 140V iGPU), 32 GB RAM
- Engines used: `qwen` (Qwen3-TTS 0.6B Base, cloned profiles) and `qwen_custom_voice` (0.6B), Japanese text, via REST API (`POST /generate`)

## Summary

`voicebox-server.exe` RSS grows with every generation and is never released. After roughly 15–20 generations the process reaches 6–12 GB, at which point **new generations never finish** — they stay `"generating"` indefinitely (observed > 60 minutes for a 6-character text that normally takes ~45 s). Every generation queued after the first stuck one also hangs.

Observed RSS progression (same process, no restart):

| point | voicebox-server RSS |
|---|---|
| fresh start | ~350 MB |
| after ~15 generations | 6.7 GB |
| after ~25 generations | 12.2 GB |

## Steps to reproduce

1. Fresh app start, download `qwen-tts-0.6B` / `qwen-custom-voice-0.6B`.
2. Via REST API, run consecutive Japanese generations (30–60 chars each) with a cloned or preset profile — e.g. 6 texts per batch, several batches back-to-back.
3. Watch `voicebox-server.exe` RSS grow monotonically in Task Manager.
4. After RSS reaches several GB, submit another generation.

## Actual result

- The generation stays `status: "generating"` forever (no error, no timeout). `GET /generate/{id}/status` returns empty; `GET /history/{id}` shows `generating` with no error.
- All subsequently queued generations also hang.
- `POST /generate/{id}/cancel` works (marks them failed), and after cancelling, often exactly **one** next generation succeeds before the hang returns.
- Killing and restarting the app fully restores normal behavior (~45–90 s per generation) — until memory grows again.

## Expected

- Memory released between generations (or a bounded cache), and/or
- A per-generation timeout that fails the job instead of wedging the whole queue.

## Notes

- Reproduced consistently across 3 separate app restarts in one session.
- First-generation-after-cancel succeeding suggests the queue worker itself survives, and the hang correlates with memory pressure rather than with any specific text/profile (same text+profile that hung completes in ~45 s on a fresh process).
- Happy to provide more logs/timing data if useful.

**Comment by gtrusler** [comment]
Confirming this on **Apple Silicon / macOS** — the issue currently only documents Windows 11 + CPU backend, so this appears to be platform-independent rather than specific to the CPU backend.

**Environment**
- Voicebox 0.5.0
- macOS 26.5.2 (build 25F84), Apple M4 Max, 36 GB RAM

**What I measured**

`voicebox-server` reached **6,843 MB RSS after 22.65 hours** of continuous uptime — a sustained average of **~5.0 MB/min**. That lands almost exactly on the "6.7 GB after ~15 generations" figure reported above, which supports the per-generation accumulation model rather than a purely time-based leak. My generation volume is high (Voicebox is driven by an automated notification pipeline, so it synthesizes many short utterances per hour), which would explain reaching that figure in under a day.

**Kernel-level corroboration**

This went past "app using a lot of RAM" and destabilized the host. macOS logged a `JetsamEvent` that names the process directly:

```
largestProcess: voicebox-server
```

At that moment, system-wide:

| Metric | Value |
|---|---|
| Free RAM | 398 MB of 36 GB |
| VM compressor holding | 15.66 GB |
| Uncompressed equivalent | 33.27 GB |
| Compressions / decompressions | 31,936,759 / 24,819,290 |
| Resident demand, all processes | 39.49 GB on 36 GB physical |

The machine became unresponsive and required a hard reboot. One system indexing daemon was killed outright by the memory pressure handler (`per-process-limit`). So on a memory-constrained host this leak can take the whole OS down, not just Voicebox.

**Possibly relevant: multiple concurrent server processes**

I consistently see **three** `voicebox-server` processes alive at once, with very different footprints — e.g. 1 MB, 171 MB, and 14 MB shortly after launch, with one growing into the multi-GB range over time. Is that expected? If only one should be serving, the extra processes may be a separate bug, or may be involved in whatever fails to release memory between generations.

**Workaround confirmation**

Restarting the process returns it to a ~350 MB baseline, matching what's reported above. On macOS this is straightforward to automate as a scheduled restart, which makes the leak survivable but not fixed.

Happy to capture a heap profile, `vmmap`, or `leaks` output from a grown process on Apple Silicon if that would be useful — just say what form is most helpful.

**Comment by gtrusler** [comment]
Follow-up with per-engine numbers, a correction to my own figure above, and one hypothesis I can rule out.

## The leak is per-generation, and the rate is engine-dependent

Each engine on a **freshly restarted server**, 6 warm-up generations to absorb lazy model loading, then 20 measured:

| engine | baseline | after 20 gens | growth | per generation |
|---|---|---|---|---|
| LuxTTS (voice clone) | 1,729 MB | 2,157 MB | +427 MB | **+21.4 MB** |
| Kokoro (preset) | 1,883 MB | 4,277 MB | +2,394 MB | **+119.7 MB** |

LuxTTS reproduced within 8% across two methods (+19.73 MB/gen in a same-process A/B, +21.38 MB/gen from cold). At 21.4 MB/gen, ~320 generations over 22.65 h accounts for the 6,843 MB I reported above.

Two details that might narrow the search:

- **Growth is a staircase, not a slope** — ~58 MB steps every 3-4 generations for LuxTTS, ~380 MB steps for Kokoro, with flat plateaus between. That looks more like a chunked cache or arena than a per-request allocation.
- **LuxTTS grew 427 MB while emitting ~10 MB of audio**, about 40x the data it outputs, so this is retained intermediates rather than output buffering.

## A hypothesis I can rule out

I found the unbounded tensor cache in `backend/utils/cache.py` and was confident that was the cause. **It isn't.** That cache is referenced only by the cloning backends, yet Kokoro touches it zero times and leaks 5.6x worse. Recording that explicitly so nobody else re-runs it.

That points at code shared by every engine — the generation/stream path — rather than anything clone-specific. Untested suspects on my side: `services/generation.py`, `history.py`, and the MLX decode path.

## Correction to my figure above: that was `phys_footprint`, not RSS

I labelled the 6,843 MB as RSS. It isn't, and the distinction turns out to matter a lot for anyone trying to measure this.

Under memory pressure macOS compresses a process's pages, so **RSS falls while real consumption climbs**. One live server, same PID, one moment:

| metric | value |
|---|---|
| `ps -o rss=` | 365 MB |
| `phys_footprint` | 9,018 MB |

RSS understated real memory by **24x**, and it understates worst exactly when the machine is in trouble. `phys_footprint` is also the metric jetsam kills on, so it is the one that decides whether the host survives. On macOS: `footprint -p <pid>` (no sudo, ~0.05 s even on a 5 GB process). Anyone monitoring this with `ps` will see a reassuringly flat curve while the box dies.

On the multiple processes I asked about above: only one actually serves, and it is not the one `ps | grep voicebox-server` returns first. The bystanders sit at ~1 MB and ~9 MB while the real server was at 7,205 MB. Resolving the PID from the listening socket (`lsof -nP -iTCP:<port> -sTCP:LISTEN -t`) is the reliable way.

## Possibly relevant to the "hangs forever" half of this issue

Generation latency degrades sharply as footprint grows, on the same server and identical input:

| server footprint | latency for a short utterance |
|---|---|
| ~1,986 MB (freshly warmed) | **1.0 s** |
| ~7,205 MB (after accumulation) | **19.9 s** |

So the reported hang may be the same phenomenon at the far end rather than a separate deadlock — everything just gets progressively slower as the process bloats. Worth checking whether the "stuck in generating" reports correlate with footprint rather than with any particular request.

## Workaround, refined

Restarting reclaims it fully. A freshly warmed LuxTTS server settles at ~1,986 MB, so I now recycle automatically on **growth above that warm baseline** (+1,500 MB) rather than on an absolute threshold — the baseline is just whatever the loaded model weighs, and an absolute ceiling set below it recycles on every generation. Generation count and uptime act as secondary triggers, and the recycle waits for an idle moment so no in-flight generation is severed.

Still happy to capture a heap profile, `vmmap`, or `leaks` output from a grown process on Apple Silicon — the reproduction is reliable and takes about 20 minutes per engine.

**Comment by gtrusler** [comment]
Spent some time in the 0.5.0 source against a live server. I haven't isolated the exact retaining allocation, but I can rule out three plausible causes and I think I've found why nothing is ever released on macOS — including a one-line-ish defect that explains the engine difference I posted above.

## Neither engine has a working cache-release path on macOS

`backends/base.py::empty_device_cache()` is the only cache-releasing helper in the backend:

```python
def empty_device_cache(device: str) -> None:
    import torch
    if device == "cuda" and torch.cuda.is_available():
        torch.cuda.empty_cache()
    elif device == "xpu" and hasattr(torch, "xpu"):
        torch.xpu.empty_cache()
```

There is no `mps` branch, though PyTorch ships `torch.mps.empty_cache()`.

That matters because **LuxTTS explicitly runs on MPS** — `luxtts_backend.py:40` calls `get_torch_device(allow_mps=True, allow_xpu=True)`. So `unload_model()` does:

```python
device = self.device          # "mps" on Apple Silicon
del self.model
empty_device_cache(device)    # falls through both branches — does nothing
```

Kokoro reaches the same dead end by the other road. `kokoro_backend.py:135` opts out with `get_torch_device(allow_mps=False)` (comment: *"MPS needs fallback env var"*), so it runs on CPU, and its unload does:

```python
if torch.cuda.is_available():
    torch.cuda.empty_cache()
```

which is unconditionally false on a Mac.

So on Apple Silicon **every engine's release path is a silent no-op**, by two different routes. That may be worth a look as a partial explanation for the 5.6x rate difference I measured between them — they're on different devices (MPS vs CPU) with different allocator behaviour, but neither ever gets told to release.

Also worth noting: every call site of `empty_device_cache` is on a *model-unload* path. Nothing releases anything after a generation, by design.

## Measured: `POST /models/unload` reclaims exactly zero

This is the part I'd most like a maintainer's eyes on. After 12 generations:

| step | `phys_footprint` | `/health` |
|---|---|---|
| after 12 generations | 2,494 MB | `model_loaded: true` |
| after `POST /models/unload` → HTTP 200 | **2,494 MB** | `model_loaded: false` |

The model genuinely unloaded — the health flag flipped — and not one megabyte came back. Consistent with the missing `mps` branch above: the model reference is dropped, the allocator is never asked to release, and the pages stay.

## Three hypotheses I can rule out

**The voice-prompt cache (`utils/cache.py::_memory_cache`).** I flagged this earlier and I was wrong, for a stronger reason than the one I first gave. Its key is `md5(reference_audio_bytes + reference_text)`, so a fixed profile occupies exactly **one** entry regardless of how many generations run. It cannot grow per-generation even for the cloning backends.

**The task registry (`services/task_queue.py`).** `_running_generation_tasks: dict[str, asyncio.Task]` has the right shape for a leak — a retained `Task` pins its coroutine frame — but it's popped in a `finally:` and the fire-and-forget set uses `add_done_callback(_background_tasks.discard)`. Clean.

**Thread-pool arenas.** Every backend generates via `asyncio.to_thread`, and per-thread allocator arenas would produce exactly the staircase this leak shows. Measured: thread count goes 19 → 21 on the first generation then **stays flat at 21** while footprint keeps climbing. Not threads.

## Staircase, measured on footprint

12 generations, unique text each, LuxTTS:

| gens | footprint | cumulative |
|---|---|---|
| 0 | 2,250 MB | — |
| 1-2 | 2,499 MB | +249 MB |
| 3-6 | 2,556 MB | +306 MB |
| 7-10 | 2,614 MB | +364 MB |
| 11-12 | 2,670 MB | +420 MB |

~55 MB steps every 3-4 generations with flat plateaus between. Repeating *identical* text still steps but more slowly (+248 MB over the same 12), so it isn't purely keyed on request content.

## Minor: `/generate/stream` is invisible to history

Unrelated to the leak but it cost me a measurement round — the streaming endpoint never enters the generation queue and never writes a history row, so `total_generations` in `/history/stats` doesn't move for streamed requests. Mine sat at 45 across a dozen live streaming calls. Anyone trying to correlate the leak with generation count via that endpoint will get a flat line.

## Suggested next steps

1. Add an `mps` branch to `empty_device_cache()`, and replace the hardcoded `torch.cuda.empty_cache()` in `kokoro_backend.py` with a call to it. Dead code on every Apple Silicon machine today.
2. Consider calling the release helper after generation, not only after unload.
3. `POST /models/unload` reclaiming zero bytes looks like its own bug, whether or not it's the same root cause.

What I could **not** do is isolate the specific object or arena holding the residue — that needs a heap profiler (`tracemalloc`, `memray`, `vmmap -summary` diffs) against a source build, and the shipped binary is a frozen 508 MB bundle. If someone points me at a dev-build setup I'm happy to run that and report back.

**Comment by dragonwang1019** [comment]
I can confirm this issue on macOS with v0.5.0. Observed behavior:

- Texts under ~14 words generate normally (2-5 seconds on CPU backend with Qwen TTS 1.7B)
- Texts over ~15 words hang at `status: "generating"` with `duration: 0.0` indefinitely
- The GPU shows `backend_variant: "cpu"` in the /health endpoint
- Cancelling and retrying doesn't help
- After a stuck generation, even short texts sometimes fail until Voicebox is restarted

**Workaround:** Split long text into chunks of ~14 words or fewer and generate sequentially. For very long monologues (50+ words), generate each chunk separately — they complete in ~4 seconds each with the PyTorch backend.

**Memory observation:** Before the hangs, the voicebox-server process shows ~380MB RSS. After one or two successful generations, it climbs to ~450MB. After a hang, it stays at ~500MB+.

Let me know if I can provide more debug info from the API side.
